### PR TITLE
Cisco ASR IOS XR Support

### PR DESCRIPTION
**Pull request by guptaamarjitarkin**

### CODE REVIEW

File: network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py [hunk 1]
```diff
@@ -3,9 +3,30 @@
...
+ADMIN_ST_KEY = 'administrativeStatus'
+OP_ST_KEY = 'operationalStatus'
+HW_KEY = 'hardwareAddress'
+DUPLEX_KEY = 'duplex'
+OP_SPEED_KEY = 'operationalSpeed'
+ACTIVE_PORTS_KEY = 'activePorts'
+PASSIVE_PORTS_KEY = 'passivePorts'
+VLAN_KEY = 'vlan'
+CONNECTED_KEY = 'connected'
+SW_PORT_KEY = 'switchPortMode'
```
Comment (piyushvmware) [inline network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py]: SP_MODE_KEY is that you wanted ?

File: network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py [hunk 2]
```diff
@@ -41,6 +62,289 @@ def parse_hardware_block(output_lines, lines):
...
+        parser.add_rule(Rule(PASSIVE_PORTS_KEY, passive_regex, rule_match_callback))
+        output_lines = parser.process(data)
+        if not bool(output_lines):
+            return []
+
+        for r in output_lines:
+            py_logger.info("Processing row {}".format(r))
+            r.update(duplex=r[DUPLEX_KEY].upper())
+            if r[IF_SPEED_KEY] == '':
+                r[IF_SPEED_KEY] = '0'
```
Comment (piyushvmware) [inline network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py]: is that required to set 0 ? 
Or can be upload csv with empty values ?

File: network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py (hunk 1 above)
Comment (piyushvmware) [inline network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py]: Can we move all these keys to common place ? Then we can reuse. Check if there is a constant python file.

File: network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py (hunk 2 above)
Comment (guptaamarjitarkin) [reply network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py]: It's needed as we are performing mathematical computation over it.

File: network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py (hunk 1 above)
Comment (guptaamarjitarkin) [reply network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py]: Will do.

Comment (guptaamarjitarkin) [reply network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py]: Done.

Comment (guptaamarjitarkin) [reply network_insight_sdk_generic_datasources/routers_and_switches/cisco/cisco_pre_post_processor.py]: Done.

**Comment by guptaamarjitarkin** [comment]
Based on discussions with @piyushvmware, we will move the file input approach to a different PR. Hence, reverted the same in current PR.

File: network_insight_sdk_generic_datasources/parsers/text/text_processor.py [hunk 3]
```diff
@@ -92,3 +111,7 @@ def rule_match_callback(**kwargs):
     field_name = kwargs['field_name']
     if groups is not None and len(groups) > 0:
         keyval[field_name] = groups[0]
+
+
+class BlockRule(Rule):
+    pass
```
[the remark this answers is not in the document]
Comment (guptaamarjitarkin) [reply network_insight_sdk_generic_datasources/parsers/text/text_processor.py]: Done.

File: network_insight_sdk_generic_datasources/parsers/text/text_processor.py [hunk 4]
```diff
@@ -48,12 +48,31 @@ def process(self, text):
...
+                            retrieve_line_count = int(list(parsed_key_values.values())[0])
+                            cnt = 1
+                            retrieve_val = ''
+                            while current_line_number + cnt < total_lines and cnt <= retrieve_line_count:
+                                retrieve_val += lines[current_line_number + cnt]
+                                if cnt != retrieve_line_count:
+                                    retrieve_val += '\n'
+                                cnt += 1
+                            parsed_key_values.update({list(parsed_key_values.keys())[0]: retrieve_val})
+                            row.update(parsed_key_values)
```
[the remark this answers is not in the document]
Comment (guptaamarjitarkin) [reply network_insight_sdk_generic_datasources/parsers/text/text_processor.py]: Done.

**Pull request merged**